<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Use case • neighborfinder</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Use case">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">neighborfinder</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/NeighborFinder_technical_report.html">Technical Report</a></li>
    <li><a class="dropdown-item" href="../articles/NeighborFinder_vignette.html">Use case</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Use case</h1>
            
      

      <div class="d-none name"><code>NeighborFinder_vignette.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/rmarkdown" class="external-link">rmarkdown</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">neighborfinder</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'neighborfinder'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:rmarkdown':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     metadata</span></span></code></pre></div>
<div class="section level2">
<h2 id="presentation-of-dataset-crc-example">Presentation of dataset: CRC Example<a class="anchor" aria-label="anchor" href="#presentation-of-dataset-crc-example"></a>
</h2>
<p>The data provided in the package was extracted from this repository:
<a href="https://doi.org/10.57745/7IVO3E" class="external-link">Taxonomic profiles, functional
profiles and manually curated metadata of human fecal metagenomes from
public projects coming from colorectal cancer studies</a></p>
<p>Three subgroups are defined following 3 geographic areas: Japan,
China and Europe (gathering Italy, Austria, Germany and France). The
focus is on patients with colorectal cancer.</p>
</div>
<div class="section level2">
<h2 id="preview-of-the-data">Preview of the data<a class="anchor" aria-label="anchor" href="#preview-of-the-data"></a>
</h2>
<p>Below, each object needed in this tutorial is loaded and a preview is
displayed.</p>
<div class="section level3">
<h3 id="the-abundance-table">1) The abundance table<a class="anchor" aria-label="anchor" href="#the-abundance-table"></a>
</h3>
<p>Here is a preview of what an abundance table from the data object
looks like, with species as rows and samples as columns. The first
column is the taxonomic annotation at the species level, the second one
is the module ID, and then each column is a sample.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span>
<span><span class="va">data</span><span class="op">$</span><span class="va">CRC_JPN</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="42%">
<col width="13%">
<col width="14%">
<col width="14%">
<col width="14%">
</colgroup>
<thead><tr class="header">
<th align="left">species</th>
<th align="left">msp_id</th>
<th align="right">DRS091121</th>
<th align="right">DRS091048</th>
<th align="right">DRS091122</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Blastocystis sp. subtype 3</td>
<td align="left">msp_0001</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
</tr>
<tr class="even">
<td align="left">Blastocystis sp. subtype 1</td>
<td align="left">msp_0002</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
</tr>
<tr class="odd">
<td align="left">Bacteroides cellulosilyticus</td>
<td align="left">msp_0003</td>
<td align="right">0e+00</td>
<td align="right">8e-07</td>
<td align="right">2e-07</td>
</tr>
<tr class="even">
<td align="left">Blastocystis sp. subtype 2</td>
<td align="left">msp_0004</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
<td align="right">0e+00</td>
</tr>
<tr class="odd">
<td align="left">Escherichia coli</td>
<td align="left">msp_0005</td>
<td align="right">1e-07</td>
<td align="right">9e-07</td>
<td align="right">0e+00</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="the-metadata">2) The metadata<a class="anchor" aria-label="anchor" href="#the-metadata"></a>
</h3>
<p>The metadata gathers characteristics (in columns) for each sample (in
rows).</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span>
<span><span class="va">metadata</span><span class="op">$</span><span class="va">CRC_JPN</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="10%">
<col width="21%">
<col width="23%">
<col width="18%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="right">HQ_clean_read_count</th>
<th align="right">gut_mapped_read_count</th>
<th align="left">gut_mapped_pc</th>
<th align="right">oral_mapped_read_count</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">DRS091371</td>
<td align="right">39159954</td>
<td align="right">32390300</td>
<td align="left">82.7128142183211</td>
<td align="right">32390300</td>
</tr>
<tr class="even">
<td align="left">DRS091370</td>
<td align="right">35443586</td>
<td align="right">28999607</td>
<td align="left">81.8190546520885</td>
<td align="right">28999607</td>
</tr>
<tr class="odd">
<td align="left">DRS091369</td>
<td align="right">43922552</td>
<td align="right">36841645</td>
<td align="left">83.8786530436574</td>
<td align="right">36841645</td>
</tr>
<tr class="even">
<td align="left">DRS091368</td>
<td align="right">29627016</td>
<td align="right">25334944</td>
<td align="left">85.5129790998864</td>
<td align="right">25334944</td>
</tr>
<tr class="odd">
<td align="left">DRS091366</td>
<td align="right">49606076</td>
<td align="right">41361270</td>
<td align="left">83.3794432762632</td>
<td align="right">41361270</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="the-taxonomy">3) The taxonomy<a class="anchor" aria-label="anchor" href="#the-taxonomy"></a>
</h3>
<p>The taxonomic file indicates for each msp (<em>i.e.</em> metagenomic
species) its species and genus annotation as well as the catalog in
which the msp is mostly found.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="va">taxo</span><span class="op">)</span></span>
<span><span class="va">taxo</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">msp_id</th>
<th align="left">species</th>
<th align="left">genus</th>
<th align="left">catalogue</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">msp_0001</td>
<td align="left">Blastocystis sp. subtype 3</td>
<td align="left">Blastocystis</td>
<td align="left">gut</td>
</tr>
<tr class="even">
<td align="left">msp_0002</td>
<td align="left">Blastocystis sp. subtype 1</td>
<td align="left">Blastocystis</td>
<td align="left">gut</td>
</tr>
<tr class="odd">
<td align="left">msp_0003</td>
<td align="left">Bacteroides cellulosilyticus</td>
<td align="left">Bacteroides</td>
<td align="left">gut</td>
</tr>
<tr class="even">
<td align="left">msp_0004</td>
<td align="left">Blastocystis sp. subtype 2</td>
<td align="left">Blastocystis</td>
<td align="left">gut</td>
</tr>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">Escherichia coli</td>
<td align="left">Escherichia</td>
<td align="left">gut &amp; oral</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="the-graph">4) The graph<a class="anchor" aria-label="anchor" href="#the-graph"></a>
</h3>
<p>This dataframe is an adjacency matrix encoding the graph with
“cluster-like” structure produced with <code><a href="../reference/graph_step.html">graph_step()</a></code>. A
value of 1 corresponds to an edge. This object is needed to simulate
semi-synthetic data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/data.html">data</a></span><span class="op">(</span><span class="va">graphs</span><span class="op">)</span></span>
<span><span class="va">graphs</span><span class="op">$</span><span class="va">CRC_JPN</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">msp_0003</th>
<th align="right">msp_0005</th>
<th align="right">msp_0007</th>
<th align="right">msp_0008</th>
<th align="right">msp_0009</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">msp_0003</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">msp_0005</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">msp_0007</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">msp_0008</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">msp_0009</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">graphs</span><span class="op">$</span><span class="va">CRC_JPN</span><span class="op">[</span><span class="fl">5</span><span class="op">:</span><span class="fl">10</span>, <span class="fl">87</span><span class="op">:</span><span class="fl">90</span><span class="op">]</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left"></th>
<th align="right">msp_0142</th>
<th align="right">msp_0145</th>
<th align="right">msp_0146c</th>
<th align="right">msp_0147</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">msp_0009</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">msp_0010</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">msp_0011</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">msp_0012</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td align="left">msp_0013</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td align="left">msp_0014</td>
<td align="right">0</td>
<td align="right">1</td>
<td align="right">0</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="aim-of-this-use-case">Aim of this use case<a class="anchor" aria-label="anchor" href="#aim-of-this-use-case"></a>
</h2>
<p>In this example, we are focusing on <em>Escherichia coli</em> because
it is one bacterium that is associated with colorectal cancer (CRC).
Here is the <a href="https://doi.org/10.2174/1389201022666210910094827" class="external-link">article</a>
mentionning it. Some <em>E. coli</em> strains can produce colibactin,
which is a toxin inducing DNA damage that may lead to colorectal
cancer.</p>
<p>The goal here is to explore the ecosystem around <em>E. coli</em> by
identifying its direct neighbors in patients with colorectal cancer.</p>
</div>
<div class="section level2">
<h2 id="test-if-the-default-parameters-of-neighborfinder-are-suitable-for-your-species-of-interest-dataset">Test if the default parameters of NeighborFinder are suitable for
your species of interest &amp; dataset<a class="anchor" aria-label="anchor" href="#test-if-the-default-parameters-of-neighborfinder-are-suitable-for-your-species-of-interest-dataset"></a>
</h2>
<p>This function enables one to test the NeighborFinder method for
different parameter values thanks to simulated data based on the
provided dataset. There will be different performance scores depending
on the species of interest and the dataset provided.</p>
<p>To do this step, generating a graph is needed.</p>
<p>It is recommended to save it to make next steps quicker. This step
was commented since it takes few minutes to execute and also because it
is included in the ‘graphs’ object.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># G &lt;- graph_step(data_with_annotation = data$CRC_JPN_CHN_EUR,</span></span>
<span><span class="co">#                 col_module_id = "msp_id",</span></span>
<span><span class="co">#                 annotation_level = "species"</span></span>
<span><span class="co">#                 )</span></span>
<span></span>
<span><span class="va">G</span> <span class="op">&lt;-</span> <span class="va">graphs</span><span class="op">$</span><span class="va">CRC_JPN_CHN_EUR</span></span></code></pre></div>
<p>NeighborFinder uses by default the level of prevalence of 30% and
filtering the top 20%. If the results from the table rendered by
<code><a href="../reference/choose_params_values.html">choose_params_values()</a></code> indicate better performance scores
at other values of these parameters, the user can adjust the
<code>prev_level</code> and <code>filtering_top</code> parameters in the
function <code><a href="../reference/apply_NeighborFinder.html">apply_NeighborFinder()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/choose_params_values.html">choose_params_values</a></span><span class="op">(</span></span>
<span>  data_with_annotation <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">CRC_JPN</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  sample_size <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  prev_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.20</span>, <span class="fl">0.25</span>, <span class="fl">0.30</span><span class="op">)</span>,</span>
<span>  filtering_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span>, <span class="fl">30</span><span class="op">)</span>,</span>
<span>  graph_file <span class="op">=</span> <span class="va">graphs</span><span class="op">$</span><span class="va">CRC_JPN</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>filtering_top <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">filtering_top</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Defining and saving true neighbors...</span></span>
<span><span class="co">#&gt; Calculating scores...</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">prev_level</th>
<th align="right">filtering_top</th>
<th align="right">F1_before</th>
<th align="right">F1_after</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">0.20</td>
<td align="right">10</td>
<td align="right">0.0120</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="right">0.20</td>
<td align="right">20</td>
<td align="right">0.0120</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="right">0.20</td>
<td align="right">30</td>
<td align="right">0.0120</td>
<td align="right">1.00</td>
</tr>
<tr class="even">
<td align="right">0.25</td>
<td align="right">10</td>
<td align="right">0.0130</td>
<td align="right">0.67</td>
</tr>
<tr class="odd">
<td align="right">0.25</td>
<td align="right">20</td>
<td align="right">0.0130</td>
<td align="right">1.00</td>
</tr>
<tr class="even">
<td align="right">0.25</td>
<td align="right">30</td>
<td align="right">0.0130</td>
<td align="right">1.00</td>
</tr>
<tr class="odd">
<td align="right">0.30</td>
<td align="right">10</td>
<td align="right">0.0058</td>
<td align="right">0.67</td>
</tr>
<tr class="even">
<td align="right">0.30</td>
<td align="right">20</td>
<td align="right">0.0058</td>
<td align="right">1.00</td>
</tr>
<tr class="odd">
<td align="right">0.30</td>
<td align="right">30</td>
<td align="right">0.0058</td>
<td align="right">1.00</td>
</tr>
</tbody>
</table>
<p>The displayed table indicates the F1 scores (harmonic mean of
precision and recall) before and after applying the NeighborFinder
method, for each <code>prev_level</code> and <code>filtering_top</code>
parameters tested. In this example, one of the combinations enabling a
satisfying performance is <code>prev_level</code>=0.30 and
<code>filtering_top</code>=30.</p>
<p>Note that this step is optional, and a summary of the expected
performance scores (calculated on 8 semi-synthetic simulated datasets)
is shown in Fig.4 of the <a href="NeighborFinder_technical_report.html">Tech report</a>.</p>
</div>
<div class="section level2">
<h2 id="apply-neighborfinder-look-for-escherichia-coli-neighbors-in-crc-patients">Apply NeighborFinder &amp; look for <em>Escherichia coli</em>
neighbors in CRC patients<a class="anchor" aria-label="anchor" href="#apply-neighborfinder-look-for-escherichia-coli-neighbors-in-crc-patients"></a>
</h2>
<p>Running <code><a href="../reference/apply_NeighborFinder.html">apply_NeighborFinder()</a></code> renders a dataframe
gathering all neighbors of your species of interest. The required data
format in input is as follows: species are the rows and samples are the
columns. The first column must be the species name, the second is the
msps name, and each subsequent column is a sample (see above for the
abundance table preview).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># JAPAN</span></span>
<span><span class="va">res_CRC_JPN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apply_NeighborFinder.html">apply_NeighborFinder</a></span><span class="op">(</span></span>
<span>  data_with_annotation <span class="op">=</span> <span class="va">data</span><span class="op">$</span><span class="va">CRC_JPN</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  prev_level <span class="op">=</span> <span class="fl">0.30</span>,</span>
<span>  filtering_top <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="op">)</span></span>
<span><span class="va">res_CRC_JPN</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">node1</th>
<th align="left">node2</th>
<th align="right">coef</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">msp_0030</td>
<td align="right">0.0872394</td>
</tr>
<tr class="even">
<td align="left">msp_0005</td>
<td align="left">msp_0103</td>
<td align="right">0.0997908</td>
</tr>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">msp_0208</td>
<td align="right">-0.1143868</td>
</tr>
<tr class="even">
<td align="left">msp_0005</td>
<td align="left">msp_0467</td>
<td align="right">-0.0870316</td>
</tr>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">msp_2448</td>
<td align="right">0.0988908</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="visualize-the-corresponding-network">Visualize the corresponding network<a class="anchor" aria-label="anchor" href="#visualize-the-corresponding-network"></a>
</h2>
<p>Here is a visualization of NeighborFinder results.</p>
<p>It is possible to adjust the size of nodes and labels. One can also
define the color of nodes corresponding to their species of interest.
The <code>taxo_option</code> generates the same network but with species
names instead of msps ones.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize_network.html">visualize_network</a></span><span class="op">(</span></span>
<span>  <span class="va">res_CRC_JPN</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="NeighborFinder_vignette_files/figure-html/network-JPN-1.png" width="672"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/visualize_network.html">visualize_network</a></span><span class="op">(</span></span>
<span>  <span class="va">res_CRC_JPN</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  annotation_option <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="NeighborFinder_vignette_files/figure-html/network-JPN-2.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="apply-neighborfinder-with-covariate-option">Apply NeighborFinder with covariate option<a class="anchor" aria-label="anchor" href="#apply-neighborfinder-with-covariate-option"></a>
</h2>
<p>Including covariates makes it possible to correct for the effect of
external factors (<em>e.g.</em> the sequencing technology used to
generate the data, the extraction method) that can introduce bias or
variability into the observations. This allows the model to better
distinguish between what is part of the true structure in the network
and what is simply due to technical or contextual differences.</p>
<p>Here is an example of applying the same method, but giving a
covariate: in this case, the name of the study. To include a covariate,
the function would run only if the following all 3 arguments are
given:</p>
<ul>
<li><p><code>covar</code>: takes a formula or the name of the column of
the covariate in the metadata table. Note that “study_accession” is
equivalent to <code>~study_accession</code>.</p></li>
<li><p><code>meta_df</code>: the dataframe giving metadata
information.</p></li>
<li><p><code>sample_col</code>: the name of the column in metadata
indicating the sample names, it should be consistent with the colnames
of “data_with_taxo”.</p></li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># On CRC patients</span></span>
<span><span class="co"># CHINA</span></span>
<span><span class="va">res_CRC_CHN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apply_NeighborFinder.html">apply_NeighborFinder</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">CRC_CHN</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  prev_level <span class="op">=</span> <span class="fl">0.30</span>,</span>
<span>  filtering_top <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  covar <span class="op">=</span> <span class="op">~</span><span class="va">study_accession</span>,</span>
<span>  meta_df <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">CRC_CHN</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"secondary_sample_accession"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># EUROPE</span></span>
<span><span class="va">res_CRC_EUR</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apply_NeighborFinder.html">apply_NeighborFinder</a></span><span class="op">(</span></span>
<span>  <span class="va">data</span><span class="op">$</span><span class="va">CRC_EUR</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  prev_level <span class="op">=</span> <span class="fl">0.30</span>,</span>
<span>  filtering_top <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  covar <span class="op">=</span> <span class="op">~</span><span class="va">study_accession</span>,</span>
<span>  meta_df <span class="op">=</span> <span class="va">metadata</span><span class="op">$</span><span class="va">CRC_EUR</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"secondary_sample_accession"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/visualize_network.html">visualize_network</a></span><span class="op">(</span></span>
<span>  <span class="va">res_CRC_CHN</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="NeighborFinder_vignette_files/figure-html/network-covariate-1.png" width="672"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/visualize_network.html">visualize_network</a></span><span class="op">(</span></span>
<span>  <span class="va">res_CRC_EUR</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  object_of_interest <span class="op">=</span> <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="NeighborFinder_vignette_files/figure-html/network-covariate-2.png" width="672"></p>
</div>
<div class="section level2">
<h2 id="look-at-the-intersection-of-neighbors-found-in-the-3-subgroups">Look at the intersection of neighbors found in the 3 subgroups<a class="anchor" aria-label="anchor" href="#look-at-the-intersection-of-neighbors-found-in-the-3-subgroups"></a>
</h2>
<p>Here is an example of how to merge the results of several datasets
into a single network. This operation can be performed with 2 or more
datasets. One can choose the visualization <code>threshold</code>,
corresponding to the minimum number of datasets in which neighbors have
been found.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/intersections_network.html">intersections_network</a></span><span class="op">(</span></span>
<span>  res_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">res_CRC_JPN</span>, <span class="va">res_CRC_CHN</span>, <span class="va">res_CRC_EUR</span><span class="op">)</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  <span class="st">"Escherichia coli"</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  edge_label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  node_size <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  annotation_option <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="NeighborFinder_vignette_files/figure-html/intersections-network-1.png" width="768"></p>
<p>Another function used here is equivalent to the network thanks to a
summary table, indicating in which datasets intersections were found
above the set <code>threshold</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/intersections_table.html">intersections_table</a></span><span class="op">(</span></span>
<span>  res_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">res_CRC_JPN</span>, <span class="va">res_CRC_CHN</span>, <span class="va">res_CRC_EUR</span><span class="op">)</span>,</span>
<span>  threshold <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  <span class="va">taxo</span>,</span>
<span>  col_module_id <span class="op">=</span> <span class="st">"msp_id"</span>,</span>
<span>  annotation_level <span class="op">=</span> <span class="st">"species"</span>,</span>
<span>  <span class="st">"Escherichia coli"</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="9%">
<col width="17%">
<col width="9%">
<col width="25%">
<col width="11%">
<col width="14%">
<col width="11%">
</colgroup>
<thead><tr class="header">
<th align="left">node1</th>
<th align="left">module1</th>
<th align="left">node2</th>
<th align="left">module2</th>
<th align="left">datasets</th>
<th align="right">intersections</th>
<th align="right">mean_coef</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">Escherichia coli</td>
<td align="left">msp_0030</td>
<td align="left">Klebsiella pneumoniae</td>
<td align="left">n_ 1, n_ 2</td>
<td align="right">2</td>
<td align="right">0.1006331</td>
</tr>
<tr class="even">
<td align="left">msp_0005</td>
<td align="left">Escherichia coli</td>
<td align="left">msp_0103</td>
<td align="left">Clostridium_AQ innocuum</td>
<td align="left">n_ 1, n_ 2</td>
<td align="right">2</td>
<td align="right">0.0744551</td>
</tr>
<tr class="odd">
<td align="left">msp_0005</td>
<td align="left">Escherichia coli</td>
<td align="left">msp_0208</td>
<td align="left">Blautia_A faecis</td>
<td align="left">n_ 1, n_ 3</td>
<td align="right">2</td>
<td align="right">-0.1269152</td>
</tr>
</tbody>
</table>
<p>One of the resulting neighbors is <em>Klebsiella pneumoniae</em>. It
is also described in this <a href="https://doi.org/10.1016/j.toxicon.2021.04.007" class="external-link">article</a> as a
bacterium associated with CRC, producing the same toxin.</p>
<p>The article summarized the prevalence of pks positive <em>E.
coli</em> and <em>K. pneumoniae</em> in CRC patient samples. pks stands
for polyketide synthase, <em>K. pneumoniae</em> genes can synthetize a
bacterial toxin which is colibactin.</p>
<blockquote>
<p>“It seems that pks positive bacteria can induce mutation of CRC
driver genes and, therefore, pks may become a marker of CRC
carcinogenesis and therapy.” (N. Strakova, K. Korena, R Karpiskova,
2021)</p>
</blockquote>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mathilde Sola, Mahendra Mariadassou, Magali Berland.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
